### PR TITLE
Promise-based workflow is now available

### PR DESCRIPTION
**Pull request by serge-medvedev**
My proposal of promise-based workflow.
To keep the code simple, I introduced an additional dependency ([event-to-promise](https://www.npmjs.com/package/event-to-promise)).
The change seems to satisfy Derek's requirement mentioned [here](https://github.com/perezd/riemann-nodejs-client/pull/27#issuecomment-398559061).
I mirrored the tests to show that both promise-based and event-based workflows are available.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,14 +11,15 @@
   ],
   "main": "riemann",
   "engines": {
-    "node": "> 0.6.0"
+    "node": ">= 0.12"
```
Comment (perezd) [inline package.json]: Is it a good idea to require this level of support? Seems like at a min we should be supporting LTS, we should think of the customers of this library.

File: riemann.js [hunk 2]
```diff
@@ -1,6 +1,18 @@
...
+// convenience methods, how convenient.
 
-// convenience method, how convenient.
 exports.createClient = function(options, onConnect) {
   return new Client(options, onConnect);
-};
\ No newline at end of file
+};
+
+exports.createPromisifiedClient = function(options) {
```
Comment (perezd) [inline riemann.js]: What about if we used options to designate the use of promises instead of a second constructor? WDYT?

File: package.json [hunk 3]
```diff
@@ -11,14 +11,15 @@
...
   },
   "repository": {
     "type": "git",
     "url": "git://github.com/perezd/riemann-nodejs-client.git"
   },
   "dependencies": {
-    "node-protobuf": "^1.4.3"
+    "node-protobuf": "^1.4.3",
+    "event-to-promise": "^0.8.0"
```
Comment (perezd) [inline package.json]: Have you reviewed this dependency closely? What to ensure we really need to take on an additional dep for our customers. Could this be inlined?

File: package.json (hunk 1 above)
Comment (serge-medvedev) [reply package.json]: My reasoning was that we have to depend on the lowest possible version our dependencies depend on.
In this case 'node-protobuf' requires >=0.12.
Please correct me if I'm wrong.

File: package.json (hunk 3 above)
Comment (serge-medvedev) [reply package.json]: This one is the most popular among similar libs, the most 'alive' on Github, has tests and has no any additional dependencies.
Actually it consists of a single file and we could repeat its functionality but I'm not sure if it's a good thing to do.

File: riemann.js (hunk 2 above)
Comment (perezd) [reply riemann.js]: yeah, perhaps: usePromises: true ?

File: package.json (hunk 1 above)
Comment (perezd) [reply package.json]: Oh did it? OK ha.

File: package.json (hunk 3 above)
Comment (perezd) [reply package.json]: SGTM.

Comment (serge-medvedev) [reply package.json]: So, are we going to leave this dependency as is OR to copy-paste the code?
I vote for the first option.

**Comment by perezd** [comment]
Let's go with dependency

**Comment by serge-medvedev** [comment]
That seems to be it: are we ready to merge this PR?

**Comment by serge-medvedev** [comment]
A question about future plans: what do you think about deprecation the event-based workflow in favor of combined promise- and callback-based one (like in MongoDB or Elasticsearch)? We could do so in the next major release (2.0.0?), still maintaining the last version with support of event-based workflow.

**Comment by perezd** [comment]
Why limit? Emitter feels like the right paradigm normally, when using udp
what does a callback do for us?

**Comment by serge-medvedev** [comment]
> what does a callback do for us?

It will tell if the message was successfully _sent_ (no DNS issues, for example).
It seems to be not too useful but is great from the point of view of uniformity when working with both TCP and UDP.

I'm not completely against Emitter interface but I don't like those warnings about 'too many listeners and possible memory leaks' Node yields (they're seen in tests, both old and new). Doesn't our abstraction seem to be a bit leaky, if we make user add something like: `emitter.setMaxListeners(0)` to prevent those?

I don't exclude the possibility that it's just me, who doesn't understand Emitter and cannot work with it properly.

**Comment by perezd** [comment]
Sure, for better or worse, many people are relying on the existing API that we have, I wouldn't be OK with removing it. I'm OK with adding additional APIs or options for working with stuff, but not removal.

**Comment by serge-medvedev** [comment]
Got it. Let's move on.

**Comment by serge-medvedev** [comment]
And before you merged this PR...
Seems to me that I see an issue in my code: what if user doesn't want to use promises and there's an error (in a `send` call) which is supposed to be handled by `error` event?
Since I return a promise, it will end up with 'unhandled rejection error'. It's just a warning now but may lead to Node crash in the future.
Have to fix it. What about letting that flag 'options.returnPromise' to determine promise usage in all places?

**Comment by perezd** [comment]
That is fine, sadly I've merged it, so can you send another PR?

**Comment by serge-medvedev** [comment]
Of course, there it is: #29 .

**Pull request merged**